### PR TITLE
Introduce background stale AppendVec shrink mechanism

### LINKED ISSUE #9196: devnet snapshots are too slow to load, due to break accounts?
#### Problem
break.solana.com broke it

STR:
1. `solana-validator --entrypoint devnet.solana.com:8001  -l ledger -o -`

The downloaded snapshot is 1.4G and takes my node about 15 minutes to load.

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

The cluster is still vulnerable account storage exhaustion attack.

This time, the attack scenario is that create-and-forget-style accounts.
Even if we keep only single (stale) account data (136 Byte) for a slot, we consume 4,194,304 Byte; that's totally waste.
This significantly increases snapshot processing time. Especially, much is wasted for index reconstruction at loading.
That's because other than that single stale account, there are usually bunch of outdated account data in the slot/AppendVec and we must conclude that those are actually not used by computation of whole data set.

As an observation of this, the current state of devnet is like this. There are so many appendvec (60000+), which contain a few accounts (~10).

#### Summary of Changes

Based on index aliveness information, periodically shrink slots by reading up alive accounts and flush them into new appropriately-fit appendVec.
So, from the AccountDB's view, there is no difference of access pattern between Banks and this background shrinker.

#### Remarks

Surprisingly, AccountsDB design is so flexible and well-coded. To implement that feature, we only needed to add new code. All of its foundation is already just existing! :)

Fixes #9196

### CODE REVIEW

**Comment by sakridge** [comment]
Cool. I thought we might need something like this.

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -1145,13 +1222,26 @@ impl AccountsDB {
         slot: Slot,
         accounts: &[(&Pubkey, &Account)],
         hashes: &[Hash],
+    ) -> Vec<AccountInfo> {
+        let write_version = self.write_version.fetch_add(accounts.len(), Ordering::Relaxed) as u64;
```
Comment (sakridge) [inline runtime/src/accounts_db.rs:1226]: The write_version was unique before for all accounts, now you can have overlapping ones if multiple threads are calling the store. I'm not sure why this change is necessary.

Comment (ryoqun) [reply runtime/src/accounts_db.rs:1226]: Well, I thought these could be valid optimization. Because these are atomic, I think there is no possible case multiple threads could be given overlapping ranges.

Say, Thread A has 100 updates and write_version is 3, then it set 103 and uses [3..103) and write_version atomically updated to 103. Then thread b has 200 updates and write_version is 103. Then, it sets 303 and uses [103..303)....

Ouch, impl is wrong.. I forgot to substract...

Anyway, this is semantically just  batching of fetch_add(+1). So, this should be problem and can avoid too much memory contention? Am I missing something? Maybe? I just noticed and jot down not to forget. :)

File: ledger-tool/src/main.rs [hunk 2]
```diff
@@ -1027,6 +1027,10 @@ fn main() {
 
                     println!("Creating a snapshot of slot {}", bank.slot());
                     bank.squash();
+                    for slot in 0..=(snapshot_slot - 1000) {
```
Comment (mvines) [inline ledger-tool/src/main.rs:1030]: Why `-1000`?   But could we just call `bank.compact_stale_slots()` instead?

**Comment by ryoqun** [comment]
@sakridge @mvines 

Howa! Super exciting result... :)

30min => 30sec (60x speedup!) on same machine!

Before:

```
$ ls -hl aaa/snapshot-2370305-427uj8wM838ZkVSMXmj9bJVmH3v9SRWeyZqZop9fnEBj.tar.bz2 
-rw-rw-r-- 1 ubuntu ubuntu 1.4G Apr  1 12:31 aaa/snapshot-2370305-427uj8wM838ZkVSMXmj9bJVmH3v9SRWeyZqZop9fnEBj.tar.bz2

$ ../../target/release/solana-ledger-tool --ledger aaa/ verify
HzyuivuNXMHJKjM6q6BE2qBsR3etqW21BSvuJTpJFj9A
[2020-04-01T03:31:50.462526157Z INFO  solana_ledger::blockstore] Maximum open file descriptors: 65000
[2020-04-01T03:31:50.656196526Z INFO  solana_ledger::blockstore] "/home/ubuntu/solana-git/data/devnet/aaa/rocksdb" open took 193ms
[2020-04-01T03:31:50.656503203Z INFO  solana_ledger::bank_forks_utils] Initializing snapshot path: "/home/ubuntu/solana-git/data/devnet/aaa/snapshot"
[2020-04-01T03:31:50.664349938Z INFO  solana_ledger::bank_forks_utils] Loading snapshot package: "/home/ubuntu/solana-git/data/devnet/aaa/snapshot-2370305-427uj8wM838ZkVSMXmj9bJVmH3v9SRWeyZqZop9fnEBj.tar.bz2"
[2020-04-01T03:38:09.662666772Z INFO  solana_ledger::snapshot_utils] snapshot untar took 379.0s
[2020-04-01T03:38:09.663310445Z INFO  solana_ledger::snapshot_utils] snapshot version: 1.1.0
[2020-04-01T03:38:09.663464681Z INFO  solana_ledger::snapshot_utils] Loading bank from "/home/ubuntu/solana-git/data/devnet/aaa/snapshot/.tmpOZkfqZ/snapshots/2370305/2370305"
[2020-04-01T03:38:09.747707612Z INFO  solana_ledger::snapshot_utils] Rebuilding accounts...
[2020-04-01T03:48:47.982082185Z INFO  solana_ledger::snapshot_utils] Rebuilding status cache...
[2020-04-01T03:48:47.984962648Z INFO  solana_ledger::snapshot_utils] Loaded bank for slot: 2370305
[2020-04-01T03:48:47.986661697Z INFO  solana_runtime::accounts_db] total_stores: 63485, newest_slot: 2370305, oldest_slot: 0, max_slot: 2261847 (num=9), min_slot: 2282647 (num=1)
[2020-04-01T03:48:48.043859433Z INFO  solana_metrics::metrics] metrics disabled: SOLANA_METRICS_CONFIG: environment variable not found
[2020-04-01T03:48:48.044950954Z INFO  solana_metrics::metrics] datapoint: accounts_db-stores total_count=63485i
[2020-04-01T03:48:49.032894572Z INFO  solana_runtime::accounts_db] scan took 331us merge took 268ns accumulate took 74us
[2020-04-01T03:48:49.032947329Z INFO  solana_runtime::bank] bank frozen: 2370305 hash: E9BoGvpsT3GpGNK2hFUuMy75sURg1huEFPEhCLCZQnow accounts_delta: 5R79LHSZLjgwThzp9SHkS376HZtQpnuPHCdah3c2VjHq signature_count: 2 last_blockhash: B6ifjTdvoneekGV7WNf9tihgka1EmnHWQYyuFmn1z9uv
[2020-04-01T03:48:49.032970401Z INFO  solana_runtime::bank] accounts hash slot: 2370305 stats: BankHashStats { num_removed_accounts: 10, num_added_accounts: 0, num_lamports_stored: 3826912914292387, total_data_len: 165103, num_executable_accounts: 0 }
[2020-04-01T03:48:49.033130186Z INFO  solana_ledger::snapshot_utils] bank rebuild from snapshot took 639.4s
[2020-04-01T03:48:49.036482638Z INFO  solana_ledger::blockstore_processor] processing ledger from slot 2370305...
[2020-04-01T03:48:49.053053081Z INFO  solana_ledger::blockstore_processor] ledger processed in 15ms. 1 MB allocated. 1 fork at 2370305, with 1 frozen bank
Ok


```

After (passed through `ledger-tool create-snapshot`)

```
$ ls -hl newnew/snapshot-2370305-427uj8wM838ZkVSMXmj9bJVmH3v9SRWeyZqZop9fnEBj.tar.bz2 
-rw-rw-r-- 1 ubuntu ubuntu 28M Apr  1 22:55 newnew/snapshot-2370305-427uj8wM838ZkVSMXmj9bJVmH3v9SRWeyZqZop9fnEBj.tar.bz2

$ ../../target/release/solana-ledger-tool --ledger newnew verify 
HzyuivuNXMHJKjM6q6BE2qBsR3etqW21BSvuJTpJFj9A
[2020-04-01T15:06:16.069683039Z INFO  solana_ledger::blockstore] Maximum open file descriptors: 65000
[2020-04-01T15:06:16.240198128Z INFO  solana_ledger::blockstore] "/home/ubuntu/solana-git/data/solana-git/data/devnet/newnew/rocksdb" open took 170ms
[2020-04-01T15:06:16.240262519Z INFO  solana_ledger::bank_forks_utils] Initializing snapshot path: "/home/ubuntu/solana-git/data/solana-git/data/devnet/newnew/snapshot"
[2020-04-01T15:06:16.241496012Z INFO  solana_ledger::bank_forks_utils] Loading snapshot package: "/home/ubuntu/solana-git/data/solana-git/data/devnet/newnew/snapshot-2370305-427uj8wM838ZkVSMXmj9bJVmH3v9SRWeyZqZop9fnEBj.tar.bz2"
[2020-04-01T15:06:39.877122270Z INFO  solana_ledger::snapshot_utils] snapshot untar took 23.6s
[2020-04-01T15:06:39.877776459Z INFO  solana_ledger::snapshot_utils] snapshot version: 1.1.0
[2020-04-01T15:06:39.877837585Z INFO  solana_ledger::snapshot_utils] Loading bank from "/home/ubuntu/solana-git/data/solana-git/data/devnet/newnew/snapshot/.tmplPJmiu/snapshots/2370305/2370305"
[2020-04-01T15:06:39.880381763Z INFO  solana_ledger::snapshot_utils] Rebuilding accounts...
[2020-04-01T15:06:47.349025503Z INFO  solana_ledger::snapshot_utils] Rebuilding status cache...
[2020-04-01T15:06:47.350857281Z INFO  solana_ledger::snapshot_utils] Loaded bank for slot: 2370305
[2020-04-01T15:06:47.352137675Z INFO  solana_runtime::accounts_db] total_stores: 59180, newest_slot: 2370305, oldest_slot: 0, max_slot: 2369872 (num=3), min_slot: 2300061 (num=1)
[2020-04-01T15:06:47.352186432Z INFO  solana_metrics::metrics] metrics disabled: SOLANA_METRICS_CONFIG: environment variable not found
[2020-04-01T15:06:47.352442461Z INFO  solana_metrics::metrics] datapoint: accounts_db-stores total_count=59180i
[2020-04-01T15:06:47.730004515Z INFO  solana_runtime::accounts_db] scan took 174us merge took 949ns accumulate took 36us
[2020-04-01T15:06:47.730069886Z INFO  solana_runtime::bank] bank frozen: 2370305 hash: E9BoGvpsT3GpGNK2hFUuMy75sURg1huEFPEhCLCZQnow accounts_delta: 5R79LHSZLjgwThzp9SHkS376HZtQpnuPHCdah3c2VjHq signature_count: 2 last_blockhash: B6ifjTdvoneekGV7WNf9tihgka1EmnHWQYyuFmn1z9uv
[2020-04-01T15:06:47.730095727Z INFO  solana_runtime::bank] accounts hash slot: 2370305 stats: BankHashStats { num_removed_accounts: 10, num_added_accounts: 0, num_lamports_stored: 3826912914292387, total_data_len: 165103, num_executable_accounts: 0 }
[2020-04-01T15:06:47.730111674Z INFO  solana_ledger::snapshot_utils] bank rebuild from snapshot took 7.9s
[2020-04-01T15:06:47.759776403Z INFO  solana_ledger::blockstore_processor] processing ledger from slot 2370305...
[2020-04-01T15:06:47.761628273Z INFO  solana_ledger::blockstore_processor] ledger processed in 1ms. 1 MB allocated. 1 fork at 2370305, with 1 frozen bank
Ok
```

File: runtime/src/accounts_db.rs (hunk 1 above)
Comment (sakridge) [reply runtime/src/accounts_db.rs:1226]: Ah, I didn't see that it does `fetch_add(accounts.len())`. Ok, yea this should be fine, although this might be 0.1% of the time of the work in this function.

Comment (ryoqun) [reply runtime/src/accounts_db.rs:1226]: Wow, `0.1%`... sad. Didn't measured. ;)

File: ledger-tool/src/main.rs (hunk 2 above)
Comment (ryoqun) [reply ledger-tool/src/main.rs:1030]: 1000 is arbitrarily chosen. I was in haste. This pr is wip to show this acutally solves #9196. Come to think of it, snapshot don't have non-rooted slots; so no need for substraction.

bank.compact_stale_slots() does take consider of `non-rooted` slots. And, it assumes constant update of `uncleaned_roots`?

File: runtime/src/accounts_db.rs [hunk 3]
```diff
@@ -879,6 +883,79 @@ impl AccountsDB {
...
+            for (pubkey, account, hash, _size, _location) in alive_accounts {
+               accounts.push((pubkey, account)); 
+               hashes.push(*hash); 
+            }
+            let write_version = self.write_version.fetch_add(accounts.len(), Ordering::Relaxed) as u64;
+            let infos = self.store_accounts_to(next_compact_slot, &accounts, &hashes, store, write_version);
+            let reclaims = self.update_index(next_compact_slot, infos, &accounts);
+            trace!("reclaim: {}", reclaims.len());
+
+            self.handle_reclaims(&reclaims);
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:934]: After this there is bunch of empty stores, so remove them.

File: runtime/src/accounts_db.rs (hunk 1 above)
Comment (sakridge) [reply runtime/src/accounts_db.rs:1226]: Just a guess :)

File: runtime/src/accounts_db.rs [hunk 4]
```diff
@@ -879,6 +881,129 @@ impl AccountsDB {
...
+        let mut stored_accounts = vec![];
+        {
+            let storage = self.storage.read().unwrap();
+            // older than recent_root stores must be root!
+            if let Some(stores) = storage.0.get(&next_compact_slot) {
+                // bail out early if the total alive account in count_and_status is big.
+                for store in stores.values() {
+                    let mut start = 0;
+                    while let Some((account, next)) = store.accounts.get_account(start) {
+                        stored_accounts.push((
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:895]: introduce a struct?

**Comment by mvines** [comment]
devnet snapshots are now too big to even load:
```
[2020-04-02T17:06:01.931593049Z ERROR solana_metrics::metrics] datapoint: panic program="validator" thread="main" one=1i message="panicked at 'Load from snapshot failed: UnpackError(Archive(\"too large snapshot: 536875106304\"))', ledger/src/bank_forks_utils.rs:67:41" location="ledger/src/bank_forks_utils.rs:67:41"
```

**Comment by ryoqun** [comment]
@sakridge I think this pr is ready for serious code review! :)

**Comment by ryoqun** [comment]
CI is unstable...

File: runtime/src/bank.rs [hunk 5]
```diff
@@ -1959,6 +1959,7 @@ impl Bank {
     /// calculation and could shield other real accounts.
     pub fn verify_snapshot_bank(&self) -> bool {
         self.clean_accounts();
+        self.shrink_all_stale_slots();
```
Comment (sakridge) [inline runtime/src/bank.rs:1962]: What is the performance of this?

File: runtime/src/accounts_db.rs [hunk 6]
```diff
@@ -879,6 +881,144 @@ impl AccountsDB {
...
+            let mut hashes = Vec::with_capacity(alive_accounts.len());
+            let mut write_versions = Vec::with_capacity(alive_accounts.len());
+
+            for (pubkey, account, hash, _size, _location, write_version) in alive_accounts {
+                accounts.push((pubkey, account));
+                hashes.push(*hash);
+                write_versions.push(*write_version);
+            }
+
+            let shrank_store = self.create_and_insert_store(slot, aligned_total);
```
Comment (sakridge) [inline runtime/src/accounts_db.rs:968]: `shrunken_store`

File: runtime/src/accounts_db.rs [hunk 7]
```diff
@@ -2228,6 +2396,23 @@ pub mod tests {
             }
         }
 
+        fn append_vec_count_for_slot(&self, slot: Slot) -> usize {
```
Comment (sakridge) [inline runtime/src/accounts_db.rs:2399]: Isn't this the accounts count for the slot?

File: runtime/src/accounts_db.rs [hunk 8]
```diff
@@ -879,6 +881,144 @@ impl AccountsDB {
...
+                            false
+                        }
+                    },
+                )
+                .collect()
+        };
+
+        let alive_total: u64 = alive_accounts
+            .iter()
+            .fold(0, |t, (_, _, _, a, _, _)| t + *a as u64);
```
Comment (sakridge) [inline runtime/src/accounts_db.rs:945]: would prefer more descriptive variable names here

File: runtime/src/accounts_db.rs [hunk 9]
```diff
@@ -879,6 +881,144 @@ impl AccountsDB {
...
+                            account.clone_account(),
+                            *account.hash,
+                            next - start,
+                            (store.id, account.offset),
+                            account.meta.write_version,
+                        ));
+                        start = next;
+                    }
+                }
+                if (alive_count as f64 / stored_accounts.len() as f64) >= 0.80 {
```
Comment (sakridge) [inline runtime/src/accounts_db.rs:907]: I think f32 is sufficient

File: runtime/src/accounts_db.rs [hunk 10]
```diff
@@ -879,6 +881,144 @@ impl AccountsDB {
...
+
+    fn next_shrink_slot(&self) -> Option<Slot> {
+        let mut candidates = self.shrink_candidate_slots.lock().unwrap();
+        let next = candidates.pop();
+        if next.is_some() {
+            next
+        } else {
+            let mut new_all_slots = self.all_root_slots_in_index();
+            let next = new_all_slots.pop();
+            std::mem::replace(&mut *candidates, new_all_slots);
```
Comment (sakridge) [inline runtime/src/accounts_db.rs:995]: would prefer to take the shink_candiate_slot lock twice here to avoid taking two locks at once

File: runtime/src/accounts_db.rs (hunk 9 above)
Comment (ryoqun) [reply runtime/src/accounts_db.rs]: i see!

File: runtime/src/accounts_db.rs (hunk 8 above)
Comment (ryoqun) [reply runtime/src/accounts_db.rs]: oops. slipped my eye...

File: runtime/src/accounts_db.rs (hunk 7 above)
Comment (ryoqun) [reply runtime/src/accounts_db.rs]: Err, the code is correct and the fn name is confusing... I've fixed that!

File: ledger-tool/src/main.rs (hunk 2 above)
Comment (ryoqun) [reply ledger-tool/src/main.rs]: This is fixed by cleaning up!

File: runtime/src/accounts_db.rs (hunk 4 above)
Comment (ryoqun) [reply runtime/src/accounts_db.rs:898]: I'd skip. :)

File: runtime/src/accounts_db.rs (hunk 10 above)
Comment (ryoqun) [reply runtime/src/accounts_db.rs]: yeah, nice catch!

File: runtime/src/bank.rs (hunk 5 above)
Comment (ryoqun) [reply runtime/src/bank.rs:1962]: To be honest, I haven't measured it still, although I know this shrinking operation isn't trivial operation. I prioritized to land this for the poor devnet bullied by the break...

My general analysis (heh, not validated by measurements) are like this:

In both following cases, I think these performance cost is justifiable to avoid the DOS this PR is trying to address.

**the snapshot codepath (as you commented here)**: Given innocent snapshots, the shrinking process is basically no-op because almost all slots should be shrunken by the snapshot-producing validator already. The no-op means exactly that we're bailing out [rather early](https://github.com/solana-labs/solana/pull/9219/files#diff-2099c5256db4eb5975c8834af38f6456R914), after a new full-scan of all `AppendVec`s collectively. There should be no lock contention pressure because we're only user of AccountsDB at the time of snapshot load (=verification).

Given malicious snapshots, this adds linear time increase to reject them, ultimately bounded by `MAX_SNAPSHOT_ARCHIVE_UNPACKED_SIZE`. (By the way, I noticed `generate_index` is quadratic in that regard while working on this. but this is a different issue...)

**the non-snapshot codepath (as an additional remark)**: As validator continuously do this at the rate of 1 slot/100ms cycling all alive slots, it should be rather rare to do actual shrinking. So this is like the snapshot code path for indivisual slots. Even if it did, that slot is rooted (no other writer) and the shrunken slot is old by definition there should be no lock contention.

*Ultimately, this PR introduces 1 read of account data, amortized no read lock by batching under normal/healthy operation.* 

As a knob for performance tuning, we have [this magic number](https://github.com/solana-labs/solana/pull/9219/files#diff-2099c5256db4eb5975c8834af38f6456R908) to reduce the count of written accounts.

Also, we could make the before-bail-out cost even cheaper by maintaining the all account count in append vec.

Anyway, I'm landing this after my last-minute local testing as this is prioritized to land as i said and I think [you're aware](https://github.com/solana-labs/solana/pull/9219#pullrequestreview-387384059). :)

Comment (ryoqun) [reply runtime/src/bank.rs:1962]: @sakridge ^^ Enjoy you reading :) (oops, I've just forgot to mention.)

**Comment by ryoqun** [comment]
I've tested against TdS:

```
$ ../../target/release/solana-ledger-tool --ledger verify verify
4uhcVJyU9pJkvQyS88uRDiswHXSCkY3zQawwpjk2NsNY
[2020-04-04T02:47:30.325130200Z INFO  solana_ledger::blockstore] Maximum open file descriptors: 65000
[2020-04-04T02:47:30.555525020Z INFO  solana_ledger::blockstore] "/home/ubuntu/solana-git/data/solana-git/data/tds/verify/rocksdb" open took 230ms
[2020-04-04T02:47:30.555763365Z INFO  solana_ledger::bank_forks_utils] Initializing snapshot path: "/home/ubuntu/solana-git/data/solana-git/data/tds/verify/snapshot"
[2020-04-04T02:47:30.558851379Z INFO  solana_ledger::bank_forks_utils] Loading snapshot package: "/home/ubuntu/solana-git/data/solana-git/data/tds/verify/snapshot-6837698-DQF8qGgZEWSwf6sMLt1SJs7yuDJseVJf5MWmbdfDHFj1.tar.bz2"
[2020-04-04T02:47:31.255008518Z INFO  solana_ledger::snapshot_utils] snapshot untar took 695ms
[2020-04-04T02:47:31.255302153Z INFO  solana_ledger::snapshot_utils] snapshot version: 1.1.0
[2020-04-04T02:47:31.255359566Z INFO  solana_ledger::snapshot_utils] Loading bank from "/home/ubuntu/solana-git/data/solana-git/data/tds/verify/snapshot/.tmpRinv9G/snapshots/6837698/6837698"
[2020-04-04T02:47:31.265369244Z INFO  solana_ledger::snapshot_utils] Rebuilding accounts...
[2020-04-04T02:47:31.336619583Z INFO  solana_ledger::snapshot_utils] Rebuilding status cache...
[2020-04-04T02:47:31.375300807Z INFO  solana_ledger::snapshot_utils] Loaded bank for slot: 6837698
[2020-04-04T02:47:31.375636941Z INFO  solana_runtime::accounts_db] total_stores: 532, newest_slot: 6837698, oldest_slot: 0, max_slot: 0 (num=5), min_slot: 6818988 (num=1)
[2020-04-04T02:47:31.375881777Z INFO  solana_metrics::metrics] metrics disabled: SOLANA_METRICS_CONFIG: environment variable not found
[2020-04-04T02:47:31.376270626Z INFO  solana_metrics::metrics] datapoint: accounts_db-stores total_count=532i
[2020-04-04T02:47:31.390816975Z INFO  solana_runtime::accounts_db] scan took 342us merge took 56us accumulate took 179us
[2020-04-04T02:47:31.390985296Z INFO  solana_runtime::bank] bank frozen: 6837698 hash: 4Y63868EU67Gqpquuahfg73gKr4c9C9Q8erWwvKQ9pQH accounts_delta: FKRbdUzyjmdaqyFtyc9qWjmfH9VVYc2DLrL6QmLhxXHJ signature_count: 108 last_blockhash: 993XEfQSPwG7BtfEaBgLxR48parP7ozwQHiyjep98No3
[2020-04-04T02:47:31.391011254Z INFO  solana_runtime::bank] accounts hash slot: 6837698 stats: BankHashStats { num_removed_accounts: 111, num_added_accounts: 0, num_lamports_stored: 498103680198288811, total_data_len: 353826, num_executable_accounts: 0 }
[2020-04-04T02:47:31.391071637Z INFO  solana_ledger::snapshot_utils] bank rebuild from snapshot took 135ms
[2020-04-04T02:47:31.411145605Z INFO  solana_ledger::blockstore_processor] processing ledger from slot 6837698...
[2020-04-04T02:47:31.594386561Z INFO  solana_ledger::blockstore_processor] ledger processed in 182ms. 90 MB allocated. 1 fork at 6837698, with 1 frozen bank
Ok

```

It got smaller even more!

File: runtime/src/accounts_db.rs [hunk 11]
```diff
@@ -879,6 +881,150 @@ impl AccountsDB {
...
+    fn next_shrink_slot(&self) -> Option<Slot> {
+        let next = {
+            let mut candidates = self.shrink_candidate_slots.lock().unwrap();
+            candidates.pop()
+        };
+
+        if next.is_some() {
+            next
+        } else {
+            let mut new_all_slots = self.all_root_slots_in_index();
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:996]: After local testing, I've found this logic is too naive for real hostile environment. But first I'll merge this...

**Comment by ryoqun** [comment]
For devnet:

```
$ ls -lh ledger-devnet2
total 520M
drwxrwxr-x 1 ubuntu ubuntu 6.8M Apr  4 11:34 accounts
-rw-r--r-- 1 ubuntu ubuntu 121K Mar 15 09:00 genesis.bin
-rw-rw-r-- 1 ubuntu ubuntu  35K Apr  4 10:27 genesis.tar.bz2
drwxr-xr-x 1 ubuntu ubuntu  438 Apr  4 11:29 rocksdb
drwxrwxr-x 1 ubuntu ubuntu 1.3K Apr  4 11:40 snapshot
-rw-rw-r-- 1 ubuntu ubuntu 257M Apr  4 11:35 snapshot-2673105-Hvo6gTRQRwsxXfjrBAbTdrPzwWHfbMsW6U4AczdchEH2.tar.bz2
-rw-rw-r-- 1 ubuntu ubuntu 263M Apr  4 11:39 snapshot-2673405-AwKh91xryFiBR48DF7gCJGBJ357CqC3TahXUTtnZSNSg.tar.bz2
drwxrwxr-x 1 ubuntu ubuntu    0 Apr  4 10:27 tmp-genesis
$ ../../target/release/solana-ledger-tool --ledger verify verify
HzyuivuNXMHJKjM6q6BE2qBsR3etqW21BSvuJTpJFj9A
[2020-04-04T02:53:00.921271113Z INFO  solana_ledger::blockstore] Maximum open file descriptors: 65000
[2020-04-04T02:53:01.147578182Z INFO  solana_ledger::blockstore] "/home/ubuntu/solana-git/data/solana-git/data/devnet/verify/rocksdb" open took 226ms
[2020-04-04T02:53:01.147628374Z INFO  solana_ledger::bank_forks_utils] Initializing snapshot path: "/home/ubuntu/solana-git/data/solana-git/data/devnet/verify/snapshot"
[2020-04-04T02:53:01.148588707Z INFO  solana_ledger::bank_forks_utils] Loading snapshot package: "/home/ubuntu/solana-git/data/solana-git/data/devnet/verify/snapshot-2673405-AwKh91xryFiBR48DF7gCJGBJ357CqC3TahXUTtnZSNSg.tar.bz2"
[2020-04-04T02:55:23.087983668Z INFO  solana_ledger::snapshot_utils] snapshot untar took 141.9s
[2020-04-04T02:55:23.088313520Z INFO  solana_ledger::snapshot_utils] snapshot version: 1.1.0
[2020-04-04T02:55:23.088364502Z INFO  solana_ledger::snapshot_utils] Loading bank from "/home/ubuntu/solana-git/data/solana-git/data/devnet/verify/snapshot/.tmpXUHCRx/snapshots/2673405/2673405"
[2020-04-04T02:55:23.090073472Z INFO  solana_ledger::snapshot_utils] Rebuilding accounts...
[2020-04-04T02:56:45.269240473Z INFO  solana_ledger::snapshot_utils] Rebuilding status cache...
[2020-04-04T02:56:45.301578430Z INFO  solana_ledger::snapshot_utils] Loaded bank for slot: 2673405
[2020-04-04T02:56:45.306226075Z INFO  solana_runtime::accounts_db] total_stores: 233830, newest_slot: 2673405, oldest_slot: 0, max_slot: 2668242 (num=2), min_slot: 2625092 (num=1)
[2020-04-04T02:56:45.306310419Z INFO  solana_metrics::metrics] metrics disabled: SOLANA_METRICS_CONFIG: environment variable not found
[2020-04-04T02:56:45.306713838Z INFO  solana_metrics::metrics] datapoint: accounts_db-stores total_count=233830i
[2020-04-04T02:57:39.785437975Z INFO  solana_runtime::accounts_db] scan took 147us merge took 346ns accumulate took 33us
[2020-04-04T02:57:39.785487158Z INFO  solana_runtime::bank] bank frozen: 2673405 hash: GwKjFfZsTyZcxJftCUsFsZEkqSdAnSZ3Kiw4KgjCv6DU accounts_delta: FC2E7pto3dYVZcXng2AU97uT363D4qQKZubWn92FQJTn signature_count: 16 last_blockhash: AycqBRuRFfF6tRtiJaHQtLQWL4XEnnALERsUzyK6CSeA
[2020-04-04T02:57:39.785512759Z INFO  solana_runtime::bank] accounts hash slot: 2673405 stats: BankHashStats { num_removed_accounts: 23, num_added_accounts: 15, num_lamports_stored: 7364560775948036266, total_data_len: 161372, num_executable_accounts: 0 }
[2020-04-04T02:57:39.785524700Z INFO  solana_ledger::snapshot_utils] bank rebuild from snapshot took 136.7s
[2020-04-04T02:57:39.789773001Z INFO  solana_ledger::blockstore_processor] processing ledger from slot 2673405...
[2020-04-04T02:57:39.794234924Z INFO  solana_ledger::blockstore_processor] ledger processed in 3ms. 1 MB allocated. 1 fork at 2673405, with 1 frozen bank
Ok

```

**Comment by ryoqun** [comment]
Sad, it seems that I introduced a regression in this PR; For some reason, it stops to produce snapshots.... I'm investigating...

**Comment by ryoqun** [comment]
> Sad, it seems that I introduced a regression in this PR; For some reason, it stops to produce snapshots.... I'm investigating...

Really, odd; this bad reproduces without this PR....

**Comment by ryoqun** [comment]
Maybe affected by this? https://github.com/solana-labs/solana/pull/9271

Rebasing and testing again.

**Comment by ryoqun** [comment]
> Maybe affected by this? #9271
> 
> Rebasing and testing again.

I've run over a day without issues. Now, I'm really merging!

File: runtime/src/bank.rs (hunk 5 above)
Comment (sakridge) [reply runtime/src/bank.rs:1962]: Thanks
